src/messages/mod.rs

```diff
@@ -58,6 +60,67 @@ impl HopMessage {
     }
 }
 
+#[derive(Eq, PartialEq, Clone)]
+pub struct HopMessageWithSerializedMessage {
+    /// Wrapped signed message.
+    content: SignedRoutingMessage,
+    /// Serialized Message as received or sent to quic_p2p.
+    serialized_message: Bytes,
+}
+
+impl HopMessageWithSerializedMessage {
+    /// Serialize message and keep both SignedRoutingMessage and Bytes.
+    pub fn new(content: SignedRoutingMessage) -> Result<Self> {
+        let hop_msg = HopMessage::new(content)?;
+        let message = Message::Hop(hop_msg);
+        let serialized_message = to_network_bytes(&message)?;
+
+        let content = if let Message::Hop(hop_msg) = message {
+            hop_msg.content
+        } else {
+            unreachable!("Created as Hop can only match Hop.")
+        };
+
+        Ok(Self {
+            content,
+            serialized_message,
+        })
+    }
+
+    /// Create from part with serialized_message the bytes of the serialization
+    /// for `Message::Hop(HopMessage::new(content))`
+    pub fn new_with_serialized_message(
+        content: SignedRoutingMessage,
+        serialized_message: Bytes,
+    ) -> Self {
```

Comment:
The problem with this method is that there is no guarantee that the bytes actually match the content. Wonder if there is some non-convoluted way to statically enforce that. Maybe something like this:

```rust
enum MessageWithBytes {
    Hop(HopMessageWithBytes),
    Direct(SignedDirectMessage) // we don't need the bytes for this one
}

impl MessageWithBytes {
    fn from_bytes(bytes: &Bytes) -> Result<Self> {
        match from_network_bytes(bytes)? {
            Message::Hop(msg) => Self::Hop(HopMessageWithBytes {
                content: msg.content,
                bytes: bytes.clone()
            })
            Message::Direct(msg) => Self::Direct(msg),
        }
    }
}
```
?

Reply:
Yes, maybe something like this will be better.
Yes there is no static enforcement, it is up to the caller to do the right thing, will see if i can do what you suggest.